% To compile this document
% graphics.off();rm(list=ls());library('knitr');knit('shell-tutorial.Rnw');  for(i in 1:2) system('R CMD pdflatex -shell-escape shell-tutorial.tex')

%setwd("/Volumes/Macintosh Storage/Users/jbinder/Dropbox/Docs/Teaching/2014/Shell intro")


\documentclass{article}


<<style, echo=FALSE, results='asis'>>=
BiocStyle::latex()
@

<<options, include=FALSE>>=
options(digits=3, width=100, stringsAsFactors = FALSE)
opts_chunk$set(echo=TRUE,tidy=FALSE,include=TRUE,
               dev='png', fig.width = 5, fig.height = 3.5, comment = '#>  ', dpi = 300,
		cache = TRUE)
@



\title{Shell Tutorial}
\usepackage{amsmath}
\usepackage{minted}
\usepackage{natbib}
\usepackage{mathpazo}
\usepackage{framed}
%\usepackage{enumerate}
\usepackage{soul}
\usepackage{cases}
\setlength{\parindent}{0cm} 

\author{Janos Binder$^1$ \\[1em]$^1$European Molecular Biology Laboratory (EMBL),\\ Heidelberg, Germany\\
\texttt{janos.binder@embl.de}}


\begin{document}

\maketitle


\tableofcontents

\section{Getting started} \label{sec:prep}

First you the necessary softwares to work remotely. On windows you need the necessary clients such as \verb+Putty+ and the \verb+WinSCP+ program. On Linux and Mac OSX you can use \verb+ssh+:

\begin{verbatim}
> ssh server.name
\end{verbatim}

Let's get familiar with the shell. You will see something like:

\begin{verbatim}
jbinder@red:~> 
\end{verbatim}

Where we are now? The \verb+pwd+ command will tell us. Try it out! Check also \verb+man+ or \verb+info+ out.

\begin{verbatim}
man pwd
\end{verbatim}

Press \verb+q+ to exit from the manual.

\section{Setting up the project environment}

We organize our work into folders. How does it work from the console. 

In the modern operating system every user has her or his own home directory. One can reach it by:

\begin{verbatim}
cd ~
\end{verbatim}

The \verb+cd+ command changes the current directory and the \verb+~+ is a special character, which refers to the home directory. What is in this directory exactly? The following command will tell:

\begin{verbatim}
ls
\end{verbatim}

\verb+-la+ flags are widely used to list the hidden files and the details of the files. 

\begin{verbatim}
ls -la
\end{verbatim}

\section{Working with directories}

The \verb+.+ denote the current directory and \verb+..+ denote the parent directory. The classroom projects should be stored in \verb+projects+. It can be created by:

\begin{verbatim}
mkdir projects
cd projects
\end{verbatim}

We start working in the projects directory.

One can always go back to the parent directory, when necessary:
\begin{verbatim}
cd ..
ls -la
\end{verbatim}

Or combining more levels:

\begin{verbatim}
cd ../..
ls -la
\end{verbatim}

Going back to projects:

\begin{verbatim}
cd ~/projects
ls -la
\end{verbatim}

When some directories are no longer used,they need to be removed. The \verb+rmdir+ is the opposite command of \verb+mkdir+.

\begin{verbatim}
mkdir wrong_directory
ls -la
ls -la wrong_directory
rmdir wrong_directory
ls -la
\end{verbatim}

We use \verb+ls -la+ to check whether everything have gone fine.

\textbf{Task:} Create a project directory. It should be located under: \verb+~/projects/compartments+

\section{Manipulating files}

First we need the human knowledge dataset from \href{http://compartments.jensenlab.org/Downloads}{http://compartments.jensenlab.org/Downloads}. This file contains human subcellular localization information from model organisms databases (mostly from \emph{UniProtKB}). It can be be downloaded:

\begin{verbatim}
cd ~/projects/compartments
wget http://download.jensenlab.org/human_compartment_knowledge_full.tsv
ls -la
\end{verbatim}

You can also try:
\begin{verbatim}
cd ~/projects/compartments
curl http://download.jensenlab.org/human_compartment_knowledge_full.tsv > hkcf2.tsv
ls -la
\end{verbatim}

Now we try out a few file manipulation commands. Copying (try out the \verb+TAB+ button when typing):

\begin{verbatim}
cp human_compartment_knowledge_full.tsv hckf.tsv
ls -la
\end{verbatim}

Rename one of the copied files (be lazy, try to use \verb+TAB+ button again):

\begin{verbatim}
mv hckf.tsv hc.tsv
ls -la
\end{verbatim}

Delete them:

\begin{verbatim}
mv hc.tsv
ls -la
\end{verbatim}

The \verb+-v+ stands for verbose, and it is useful to see what is happening behind. It works with most of the unix commands. The \verb+*+ character matches to everything.

\begin{verbatim}
rm -v hc*.tsv
\end{verbatim}

\textbf{Task} Create a directory under \verb+~/projects/compartments/mouse+. Download the mouse knowledge dataset and rename the file to \verb+mckf.tsv+.

\subsection{Manipulating files}

Bioinformatics data comes often in text format. To get a glance, try the following commands.

\begin{verbatim}
head human_compartment_knowledge_full.tsv
tail human_compartment_knowledge_full.tsv
\end{verbatim}

The \verb+head+ command shows the beginning of the file, while \verb+tail+ shows the end of the file. Using \verb+-20+ will show the first or last 20 lines. The \verb+human_compartment_knowledge_full.tsv+ contains the following columns: ENSEMBL identifier of the gene, common protein name, cellular component GO identifier, name of the cellular component, source database, evidence code, confidence code.

The columns in \verb+tsv+ files are separated with a tab character (\verb+\t+). These files are closely related to the comma separated files (\verb+csv+).

Let see what is inside:

\begin{verbatim}
less human_compartment_knowledge_full.tsv
\end{verbatim}

You can use the cursors, the \verb+space+ and \verb+G+ and \verb+g+ button, which goes to the end and the beginning of the file respectively. Use \verb+q+ to exit the program. We need some statistics about the file, the \verb+wc+ command answers how many lines, words and characters exist. We can limit it further by using the flags (\verb+-l -w+ or \verb+c-+).

\begin{verbatim}
wc human_compartment_knowledge_full.tsv
\end{verbatim}

We can also filter specific lines, for example what proteins are localized in the mitochondria. The Gene Ontology identifier for mitochondrion is \verb+GO:0005739+. It can be done by:

\begin{verbatim}
grep 'GO:0005739' human_compartment_knowledge_full.tsv
\end{verbatim}

It seems to be too long, let us see a sample from the output. The pipes are useful in Unix for this purpose. For this we send the output of the command into another command.

\begin{verbatim}
grep 'GO:0005739' human_compartment_knowledge_full.tsv | head
\end{verbatim}

If it seems good, we can store the mitochondrial proteins. For that purpose we will use redirection.

\begin{verbatim}
grep 'GO:0005739' human_compartment_knowledge_full.tsv > mitochondrial_proteins.tsv
\end{verbatim}

Let us ignore annotations about NR3C1 protein. In the \verb+grep+ command using \verb+-v+ flag lists all lines, which do not contain the pattern. One example:

\begin{verbatim}
grep -v 'NR3C1' human_compartment_knowledge_full.tsv | head
\end{verbatim}

\textbf{Task} Ignore annotations which comes from the \emph{Reactome} database. Store the results in a file.

Specific columns can be also selected. Now we see how a file with pair of Ensembl identifiers and GO terms can be created. To get a feeling about the structure of the file again we use the \verb+head+ command and the \verb+cut+ command will select the respective columns.

\begin{verbatim}
head human_compartment_knowledge_full.tsv
cut -f1,3 human_compartment_knowledge_full.tsv | head
\end{verbatim}

If works then let us store the results:

\begin{verbatim}
cut -f1,3 human_compartment_knowledge_full.tsv > ensembl_go_pairs.tsv
\end{verbatim}

Where does the data come from? Which databases provided the information? Selecting the fifth column answers this question:

\begin{verbatim}
cut -f5 human_compartment_knowledge_full.tsv
\end{verbatim}

However the results should be sorted and the duplicates should be removed. \verb+sort+ orders the databases in an alphabetical order, and the \verb+-u+ flag removes the duplicates.

\begin{verbatim}
cut -f5 human_compartment_knowledge_full.tsv | sort -u
\end{verbatim}

We are also interested how many entries are coming from the various databases. Counting using \verb+uniq -c+ will answer the question.

\begin{verbatim}
cut -f5 human_compartment_knowledge_full.tsv | sort | uniq -c
\end{verbatim}

\textbf{Task} List the various evidence codes and count how many proteins are associated with that evidence code.

\subsection{Advanced operations with AWK}

AWK is an interpreted programming language designed for text processing and typically used as a data extraction and reporting tool. It is a standard feature of most Unix-like operating systems. Let us see some examples.

\subsubsection*{Printing specific columns with AWK}

As the first exercise let us see how can we print the columns with AWK. We use the GNU variant of AWK.

\begin{verbatim}
gawk -F '\t' '{print $1 "\t" $3;}' human_compartment_knowledge_full.tsv | head
\end{verbatim}

The \verb+-F '\t'+ expresses the separator between the columns. In the curly backets we express what we want to do with the column. Using print statement, we can select the columns. It says print the first and the third columns separated by a tab.

\subsubsection*{Filtering on specific values}

We can also select the mitochondrial proteins with AWK:

\begin{verbatim}
gawk -F '\t' '$3 == "GO:0005739"' human_compartment_knowledge_full.tsv | head
\end{verbatim}

Here \verb+$3 == "GO:0005739"+ is a logical expression and it says execute only if the third column is equal with that value. The opposite expression would be \verb+$3 == "GO:0005739"+. There are no curly backets used, and AWK by default will print out the lines, where the condition is true. We can combine it with the previous example.

\begin{verbatim}
gawk -F '\t' '($3 == "GO:0005739"){print $1 "\t" $3;}' human_compartment_knowledge_full.tsv | head
\end{verbatim}

\subsubsection*{More filtering possibilities}

It is often neccesary to filter a file based on some numerical threshold. In the example the last column denotes how reliable is the localization evidence. Filtering on high confidence scores ($= 5$):

\begin{verbatim}
gawk -F '\t' '$7 > 4' human_compartment_knowledge_full.tsv | head
\end{verbatim}

There are a few logical expressions, which can be used with numbers. These are \verb+==, <, >, <=, >=+.

One can combine multiple conditionals. Here we examine the mitochondrial proteins with highly reliable evidence.

\begin{verbatim}
gawk -F '\t' '$3 == "GO:0005739" && $7 > 4' human_compartment_knowledge_full.tsv | head
\end{verbatim}

The \verb+&&+ denotes that both statement needs to be true, while \verb+||+ denote that at least one statement needs be true. Let us store these resuls.

\begin{verbatim}
gawk -F '\t' '$3 == "GO:0005739" && $7 > 4' human_compartment_knowledge_full.tsv > mitochondrial_proteins_with_high_scores.tsv
\end{verbatim}

\textbf{Task 1} Print mitochondrial proteins which comes from the \emph{UniProtKB} database.

\textbf{Task 2} Print proteins which do not come from \emph{Reactome}, but it has atleast a confidence score of 4. The format should be ENSEMBL identifiers and GO identifier pairs (e.g. get rid of the unnecessary colums).

\section{Shell programming}

We will write a few script during the second part. First we learn how editing works remotely.

\subsection*{Coding the first script -- Hello BioWorld!}

\begin{verbatim}
pico 1_hello_bioworld.sh
\end{verbatim}

Write the following lines:

\begin{minted}{sh}
#!/bin/sh

# we can also have comments in the examples

echo "Hello BioWorld!"
\end{minted}

Exit works with \verb!CTRL+X!. Don't forget to save!

The \verb+#!/bin/sh+ defines that the script is a shell script. The \verb+echo+ prints to the console.

You have make the script runnable and then you can test it. On Unix every file has user, group and everybody permission levels. When you create a file you as the user can read and write it, but you cannot execute it, that is the reason why we use \verb+chmod+ to change the permission. \verb+ls -la+ also lists the permissions.

\begin{verbatim}
ls -la 1_hello_bioworld.sh
chmod a+x 1_hello_bioworld.sh
./1_hello_bioworld.sh
ls -la 1_hello_bioworld.sh
\end{verbatim}

\subsection*{Using variables}

Variables are useful to store intermediate results and often used text and numbers. 

\begin{minted}{sh}
#!/bin/bash

HW="Hello World!"

echo $HW
echo "It is boring to print" $HW
\end{minted}

When we refer to the variables we need to use the \verb+$+ sign.

\subsection*{Using conditionals}

A branch of a script is executed only if the statement is true. A simple example:

\begin{minted}{sh}
#!/bin/bash

FILE="3_basic_conditionals.sh"

if [ -f $FILE ]; then
    echo "FILE:" $FILE "exists!"
fi
\end{minted}

A more complicated example:

\begin{minted}{sh}
#!/bin/bash

FILE="3a_complex_conditionals"

if [ -f $FILE ]; then
    echo "FILE:" $FILE "exists!"
else
    echo "FILE:" $FILE "does not exists!"
fi
\end{minted}

\textbf{Task:} Try to modify this script to run the other branch of the conditional.

\subsection*{Using loops}

A loop is a sequence of statements which is specified once but which may be carried out several times in succession. The code "inside" the loop is obeyed a specified number of times, or once for each of a collection of items, or until some condition is met, or indefinitely.

\begin{minted}{sh}
#!/bin/sh

echo "We count to ten!"

for (( i = 1 ; i <= 10 ; i++ )) ; do
    echo $i
done
\end{minted}

The double bracket denote that \verb+i+ refers to a number.

\textbf{Task:} Modify the script to count back from 10.

There is another solution to the previous example with the \verb+while+ keyword.

\begin{minted}{sh}
#!/bin/sh

echo "We count to ten!"

(( i = 1 ))
while (( i <= 10 )) ; do
    echo $i
    (( i++ ))
done
\end{minted}

\textbf{Task:} Modify the script to count back from 10.

The shell can also store the output of a command. Here is an example, which writes the files out:

\begin{minted}{sh}
#!/bin/sh

FILES=`ls -1`

for f in $FILES; do
    echo $f "is a file"
done
\end{minted}
\end{document}