\documentclass{article}
\usepackage{minted}

\title{NGS Tutorial}
\usepackage{amsmath}
\usepackage{minted}
\usepackage{natbib}
\usepackage{mathpazo}
%\usepackage{enumerate}
\usepackage{soul}
\usepackage{cases}
\setlength{\parindent}{0cm} 

\author{Roland Krause$^1$ \\[1em]Luxembourg Centre for Systems Biomedicine (LCSB),\\ Unversity of Luxembourg\\
\texttt{$^1$roland.krause@uni.lu}}


\usepackage{url}
\begin{document}
\maketitle

\tableofcontents

\section{Introduction}

This part of the work shop shows how to align sequences,
improve the alignment and call variants. In order to speed up the
lecture, only chromosome 22 is taken into account.


Create directory ngs for all 
\begin{minted}{bash}
 	mkdir workshop
	chdir workshop
\end{minted}

All data source data is kept in the directory \verb+/Users/roland.krause/Public/isb101/+. 

For your convenience, create a variable holding the path to the resources.

\begin{minted}{bash}
RESOURCE="/Users/roland.krause/Public/isb101/"
\end{minted}

The programs samtools, bwa, picard-tools should be available from your path.

Question: How do you find out where a program is installed?


\section{Extracting reads from an existing BAM file}
\subsection{Copy the BAM file to your local folder}
This file has already been processed. We use it as source of our data.
The BAM file is called  \verb+AID16385_SID15899_daughter.Improved.bam+

Create a i{\em soft link} to the file.

\begin{minted}{bash}
ln -s $RESOURCE/AID16385_SID15899_daughter.Improved.bam .
\end{minted} //$
% $
Question: Why not don't we copy the file? Check the properties of the file using options of the \verb+ls+ command.

\subsection{Index the BAM file}
\begin{minted}{bash}
samtools index AID16385_SID15899_daughter.Improved.bam
\end{minted}
Question: What did the command do? What is an "index"?

\subsection{Extract chromosome 22 from the example BAM}
make a slice of chromosome 22 and save in SAM format
\begin{minted}{bash}
samtools view AID16385_SID15899_daughter.Improved.bam 22 \
> AID16385_SID15899_daughter.Improved.22.sam
\end{minted}

\subsection{Convert SAM to FASTQ using PICARD}
\begin{minted}{bash}
java -jar picard-tools-1.108/SamToFastq.jar \
 I=AID16385_SID15899_daughter.Improved.22.sam \
 F=AID16385_SID15899_daughter.22.1.fq \
 F2=AID16385_SID15899_daughter.22.2.fq \
 VALIDATION_STRINGENCY=SILENT
\end{minted}

\section{Performing quality control of the sequenced reads}
to run fastqc on a remote machine login via ssh with -X for X11 support!
the following command opens the FastQC GUI
\begin{minted}{bash}
perl FastQC/fastqc
\end{minted}

%%% goto File->Open


\section{Mapping}
\subsection{Indexing the reference }
The following command has to be use. This step is skipped as it takes to much time.
\begin{minted}{bash}
# bwa index -a bwtsw human_g1k_v37_Ensembl_MT_66.fasta
\end{minted}



\subsection{Perform alignment with bwa} 

%%% -n 7 is used for a more sensitive alignment
%%% -q 15 is used for trimming sequences by quality (default=0=switched off)
\begin{minted}{bash}
bwa mem -n 7 -q 15 human_g1k_v37_Ensembl_MT_66.fasta AID16385_SID15899_daughter.22.1.fq > AID16385_SID15899_daughter.22.1.sai AID16385_SID15899_daughter.22.2.sai 
bwa-0.7.7/bwa sampe -r '@RG\tID:FCC189PACXX:2:1101\tPL:Illumina\tLB:LIBID15900\tSM:SID15899\tCN:CCG' \
  human_g1k_v37_Ensembl_MT_66.fasta \
  AID16385_SID15899_daughter.22.1.sai \
  AID16385_SID15899_daughter.22.2.sai \
  AID16385_SID15899_daughter.22.1.fq \
  AID16385_SID15899_daughter.22.2.fq \
  > AID16385_SID15899_daughter.22.sam
\end{minted}


\subsection{Convert SAM to BAM}
\begin{minted}{bash}
samtools view -bS AID16385_SID15899_daughter.22.sam > AID16385_SID15899_daughter.22.bam
\end{minted}
\subsection{Sort BAM}
The suffix bam is automatically attached. This is for compatibility with PICARD and GATK.

\begin{minted}{bash}
samtools sort AID16385_SID15899_daughter.22.bam  AID16385_SID15899_daughter.22.sorted 
\end{minted}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Mark duplicate reads}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Create a temporary folder and run picard tools.

Copy picard tools from the \verb+RESOURCE+ folder.
\begin{minted}{bash}
mkdir tmp

java -Djava.io.tmpdir=tmp -jar picard-tools-1.108/MarkDuplicates.jar \
 I=AID16385_SID15899_daughter.22.sorted.bam \
 O=AID16385_SID15899_daughter.22.sorted.marked.bam \
 METRICS_FILE=AID16385_SID15899_daughter.22.sorted.marked.metrics \
 VALIDATION_STRINGENCY=LENIENT

\end{minted}        
%%% useful links 
\url{http://picard.sourceforge.net/command-line-overview.shtml}
%MarkDuplicates
\url{http://sourceforge.net/apps/mediawiki/picard/index.php?title=Main_Page}
%Q:_How_does_MarkDuplicates_work.3F

\subsection{Index BAM file for GATK}
\begin{minted}{bash}
samtools index AID16385_SID15899_daughter.22.sorted.marked.bam
\end{minted}


\section{BQSR(Base Quality Score Recalibration)}

Introduce GATK, dbSNP

Stage 1
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -I AID16385_SID15899_daughter.22.sorted.marked.bam \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -knownSites dbsnp_135.b37.vcf \
 -T  CountCovariates \
 -cov ReadGroupCovariate \
 -cov QualityScoreCovariate \
 -cov CycleCovariate \
 -cov DinucCovariate \
 -recalFile  AID16385_SID15899_daughter.22.sorted.marked.recal_data.csv 
\end{minted}
Stage 2
\begin{minted}{bash}
 
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -I AID16385_SID15899_daughter.22.sorted.marked.bam \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -o  AID16385_SID15899_daughter.22.sorted.marked.recal.bam \
 -T TableRecalibration \
 -recalFile AID16385_SID15899_daughter.22.sorted.marked.recal_data.csv \
 -noOQs
 \end{minted}
 

\subsection{Realign sequences close to indels}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -T RealignerTargetCreator \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -known Mills_and_1000G_gold_standard.indels.b37.sites.vcf \
 -known 1000G_phase1.indels.b37.vcf \
 -o AID16385_SID15899_daughter.22.sorted.marked.recal.bam.list \
 -I AID16385_SID15899_daughter.22.sorted.marked.recal.bam \
 -L 22
 \end{minted}

Stage 2
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar -T IndelRealigner \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -o AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.bam \
 -I AID16385_SID15899_daughter.22.sorted.marked.recal.bam \
 -targetIntervals AID16385_SID15899_daughter.22.sorted.marked.recal.bam.list
\end{minted}

\subsection{Index the bam file}
\begin{minted}{bash}
samtools index AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.bam
\end{minted}


%%%%%%%%%%%%%%%%%%%%
\section{Variant calling}
%%%%%%%%%%%%%%%%%%%%


\subsection{Samtools mpileup}
\begin{minted}{bash}
samtools-0.1.19/samtools mpileup \
 -S -E -g -Q 13 -q 20 \
 -f human_g1k_v37_Ensembl_MT_66.fasta \
 AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.bam | \
 samtools-0.1.19/bcftools/bcftools \
 view -vc - > AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.mpileup.vcf
\end{minted}

\subsection{ GATK Unified Genotyper}
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -l INFO \
 -T UnifiedGenotyper \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -I AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.bam \
 -stand_call_conf 30.0 \
 -stand_emit_conf 10.0 \
 --genotype_likelihoods_model BOTH \
 --min_base_quality_score 13 \
 --max_alternate_alleles 3  \
 -A MappingQualityRankSumTest \
 -A AlleleBalance \
 -A BaseCounts \
 -A ChromosomeCounts \
 -A QualByDepth \
 -A ReadPosRankSumTest \
 -A MappingQualityZeroBySample \
 -A HaplotypeScore \
 -A LowMQ \
 -A RMSMappingQuality \
 -A BaseQualityRankSumTest \
 -L 22 \
 -o AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.gatk.vcf
\end{minted} 

\subsection{ Visualize alignments with samtools tview }
\begin{minted}{bash}
samtools-0.1.19/samtools tview \
 AID16385_SID15899_daughter.22.sorted.marked.recal.realinged.bam \
 human_g1k_v37_Ensembl_MT_66.fasta
\end{minted}
Acknowledgement: Holger Thiele, Kamel Jabbari (CCG Cologne)
\end{document}
