\documentclass{article}
\usepackage{minted}

\title{NGS Tutorial}
\usepackage{amsmath}
\usepackage{minted}
\usepackage{natbib}
\usepackage{mathpazo}
%\usepackage{enumerate}
\usepackage{soul}
\usepackage{cases}
\setlength{\parindent}{0cm} 
\definecolor{bg}{rgb}{0.95,0.95,0.95}
\author{Roland Krause$^1$ \\[1em]Luxembourg Centre for Systems Biomedicine (LCSB),\\ Unversity of Luxembourg\\
\texttt{$^1$roland.krause@uni.lu}}


\usepackage{url}
\begin{document}
\maketitle

\tableofcontents

\section{Introduction}

This part of the work shop shows how to align sequences,
improve the alignment and call variants. In order to speed up the
lecture, only chromosome 22 for an exome study is taken into account.

\subsection{Set up}
Create directory \verb+ngs+ for all next-generation sequencing tutorials.
 
\begin{minted}{bash}
 	mkdir ngs
	cd ngs
\end{minted}





All data source data is kept in the directory \verb+/Users/roland.krause/Public/isb101/+. 

For your convenience, create a variable holding the path to the resources.

\begin{minted}{bash}
RESOURCE="/Users/roland.krause/Public/isb101/"
\end{minted}

Note: Not all commands are given in full in this tutorial. 
You might need to use the commands you have learned previously.

The programs \verb+samtools+ and \verb+bwa+ should be available from your path.

Question: How do you find out where a program is installed?
% which 

\section{Extracting reads from an existing BAM file}
\subsection{Copy the BAM file to your local folder}
This file has already been processed. We use it as source of our data.
The BAM file is called  \verb+daughter.Improved.bam+

Create a {\em soft link} to the file.

\begin{minted}{bash}
ln -s $RESOURCE/daughter.Improved.bam .
\end{minted}
% $
Questions:
\begin{enumerate}
\item Why not don't we copy the file? 
\item Check the properties of the file using options of the \verb+ls+ command.
\item What happens if you would delete the link in your directory?
\item What happens if you delete the file in \mint{bash}|$RESOURCE| ?
\end{enumerate}
\subsection{Index the BAM file}
\begin{minted}{bash}
samtools index daughter.Improved.bam
\end{minted}
This will take a few seconds. 

Questions
\begin{enumerate}
\item  What did the command do? 
	\item What is an {\em index}?
\end{enumerate}i

\subsection{Extract chromosome 22 from the example BAM}
Slice chromosome 22 and save a piece in SAM format
\begin{minted}{bash}
samtools view daughter.Improved.bam 22 \
> daughter.Improved.22.sam
\end{minted}

\subsection{Convert SAM to FASTQ using PICARD}

Create a soft link to the picard-tool \verb+SamToFastq+ in your local \verb+ngs+ directory. 

Then, run the converter as follows:

\begin{minted}{bash}
java -jar SamToFastq.jar \
 I=daughter.Improved.22.sam \
 F=daughter.22.1.fq \
 F2=daughter.22.2.fq \
 VALIDATION_STRINGENCY=SILENT
\end{minted}

Inspect the output files and recapitulte the fastq-format.

\section{Performing quality control of the sequenced reads}
FastQC is a tool kit for quality performance. You will probably not be able to run 
this unless you are working from a Linux computer. 
If you are working from a Mac, you need to have X11 or XQuartz installed.

On the server login on a remote machine login via ssh with -X for X11 support.
\mint{bash}{ ssh -X username@nitro.uni.lux}
the following command opens the FastQC GUI
\begin{minted}{bash}
perl FastQC/fastqc
\end{minted}

Goto File->Open
and load the new .sam file

We will discuss this together on screen. 

\section{Mapping}
\subsection{Indexing the reference }
The following command has to be use. This step is skipped as it takes to much time.
\begin{minted}{bash}
# bwa index -a bwtsw human_g1k_v37_Ensembl_MT_66.fasta
\end{minted}



\subsection{Perform alignment with bwa} 

%%% -n 7 is used for a more sensitive alignment
%%% -q 15 is used for trimming sequences by quality (default=0=switched off)
\begin{minted}{bash}
bwa mem human_g1k_v37_Ensembl_MT_66.fasta daughter.22.1.fq daughter.22.2.fq \
  > daughter.22.sam
\end{minted}


\subsection{Convert SAM to BAM}
\begin{minted}{bash}
samtools view -bS daughter.22.sam > daughter.22.bam
\end{minted}
\subsection{Sort BAM}
The suffix bam is automatically attached. This is for compatibility with PICARD and GATK.

\begin{minted}{bash}
samtools sort daughter.22.bam  daughter.22.sorted 
\end{minted}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Mark duplicate reads}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Create a temporary folder and run picard tools.

Copy picard tools from the \verb+RESOURCE+ folder.
\begin{minted}{bash}
mkdir tmp

java -Djava.io.tmpdir=tmp -jar picard-tools-1.108/MarkDuplicates.jar \
 I=daughter.22.sorted.bam \
 O=daughter.22.sorted.marked.bam \
 METRICS_FILE=daughter.22.sorted.marked.metrics \
 VALIDATION_STRINGENCY=LENIENT

\end{minted}        
%%% useful links 
\url{http://picard.sourceforge.net/command-line-overview.shtml}
%MarkDuplicates
\url{http://sourceforge.net/apps/mediawiki/picard/index.php?title=Main_Page}
%Q:_How_does_MarkDuplicates_work.3F

\subsection{Index BAM file for GATK}
\begin{minted}{bash}
samtools index daughter.22.sorted.marked.bam
\end{minted}


\section{BQSR(Base Quality Score Recalibration)}

Introduce GATK, dbSNP

Stage 1
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -I daughter.22.sorted.marked.bam \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -knownSites dbsnp_135.b37.vcf \
 -T  CountCovariates \
 -cov ReadGroupCovariate \
 -cov QualityScoreCovariate \
 -cov CycleCovariate \
 -cov DinucCovariate \
 -recalFile  daughter.22.sorted.marked.recal_data.csv 
\end{minted}
Stage 2
\begin{minted}{bash}
 
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -I daughter.22.sorted.marked.bam \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -o  daughter.22.sorted.marked.recal.bam \
 -T TableRecalibration \
 -recalFile daughter.22.sorted.marked.recal_data.csv \
 -noOQs
 \end{minted}
 

\subsection{Realign sequences close to indels}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -T RealignerTargetCreator \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -known Mills_and_1000G_gold_standard.indels.b37.sites.vcf \
 -known 1000G_phase1.indels.b37.vcf \
 -o daughter.22.sorted.marked.recal.bam.list \
 -I daughter.22.sorted.marked.recal.bam \
 -L 22
 \end{minted}

Stage 2
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar -T IndelRealigner \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -o daughter.22.sorted.marked.recal.realinged.bam \
 -I daughter.22.sorted.marked.recal.bam \
 -targetIntervals daughter.22.sorted.marked.recal.bam.list
\end{minted}

\subsection{Index the bam file}
\begin{minted}{bash}
samtools index daughter.22.sorted.marked.recal.realinged.bam
\end{minted}


%%%%%%%%%%%%%%%%%%%%
\section{Variant calling}
%%%%%%%%%%%%%%%%%%%%


\subsection{Samtools mpileup}
\begin{minted}{bash}
samtools-0.1.19/samtools mpileup \
 -S -E -g -Q 13 -q 20 \
 -f human_g1k_v37_Ensembl_MT_66.fasta \
 daughter.22.sorted.marked.recal.realinged.bam | \
 samtools-0.1.19/bcftools/bcftools \
 view -vc - > daughter.22.sorted.marked.recal.realinged.mpileup.vcf
\end{minted}

\subsection{ GATK Unified Genotyper}
\begin{minted}{bash}
java -Djava.io.tmpdir=tmp -jar GenomeAnalysisTK-1.6-11-g3b2fab9/GenomeAnalysisTK.jar \
 -l INFO \
 -T UnifiedGenotyper \
 -R human_g1k_v37_Ensembl_MT_66.fasta \
 -I daughter.22.sorted.marked.recal.realinged.bam \
 -stand_call_conf 30.0 \
 -stand_emit_conf 10.0 \
 --genotype_likelihoods_model BOTH \
 --min_base_quality_score 13 \
 --max_alternate_alleles 3  \
 -A MappingQualityRankSumTest \
 -A AlleleBalance \
 -A BaseCounts \
 -A ChromosomeCounts \
 -A QualByDepth \
 -A ReadPosRankSumTest \
 -A MappingQualityZeroBySample \
 -A HaplotypeScore \
 -A LowMQ \
 -A RMSMappingQuality \
 -A BaseQualityRankSumTest \
 -L 22 \
 -o daughter.22.sorted.marked.recal.realinged.gatk.vcf
\end{minted} 

\subsection{ Visualize alignments with samtools tview }
\begin{minted}{bash}
samtools-0.1.19/samtools tview \
 daughter.22.sorted.marked.recal.realinged.bam \
 human_g1k_v37_Ensembl_MT_66.fasta
\end{minted}
Acknowledgement: Holger Thiele, Kamel Jabbari (CCG Cologne)
\end{document}
