---
title: "Filtering variants"
author: "Roland Krause"
date: "30 March 2021"
output: 
  unilur::tutorial_html: default
  unilur::tutorial_pdf_solution: default
  unilur::tutorial_html_solution: default
  unilur::tutorial_pdf: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, include=FALSE}
library(tidyverse)
sam_dir <- "/Users/roland.krause/Workspace/isb101/NGS/WEScourse/corpasome/variants_samtools/"
```

## Introduction
In this short introduction we will filter variants from an annotated file produced on the shell in R, a common step in genomic data analyses. The `runAnalysis.sh`script so far is using custom Perl scripts. We want to rewrite the steps in R. 
For each question, save your results in tibble.

For filtering it is recommended to use the `str_detect()` function from the *stringr* package included in the *tidyverse*.

## Getting started

Copy data from the server to your local computer. Information is supplied on the [Moodle](https://moodle.uni.lu/course/view.php?id=93#section-18).

## Filtering SNPs recessive analysis


 1. Load the file `corpasome.trio.chr22.snps.filtered.annovar.tsv` into R. Filter the columns `corpasome_dad`, `corpasome_mom`and `corpasome_daughter` with the patterns $01$, $01$, and $11$ respectively. Compare your results to the file `corpasome.trio.chr22.snps.filtered.annovar.recessive`. Why are these patterns particularly important in this case ?
 
```{r, solution = TRUE}
recessive <- read_tsv(file.path(
  #sam_dir, 
  "~/Downloads/corpasome.trio.chr22.snps.filtered.annovar.tsv"), guess_max = 5100) %>% 
  filter(corpasome_dad == "01", 
         corpasome_mom == "01", 
         corpasome_daughter == "11")


```

 
 2. Filter the columns `esp6500si_ea` and `1000g2012apr_eur` such that the values above $0.05$ in any row are removed. Be careful about missing values in this column. Empty values need to be included! Use the function `replace_na()` to convert $NA$ to $0$. Compare to the file named `corpasome.trio.chr22.snps.filtered.annovar.recessive.noncommon`.
 
```{r, solution = TRUE}
cut_at <- 0.05 

recessive %>% 
  mutate(esp = replace_na(esp6500si_ea, 0)) %>% 
  mutate(tg = replace_na(`1000g2012apr_eur`, 0)) %>% 
  filter(esp <= cut_at ,
          tg <= cut_at ) -> uncommon
```
 
 3. From the results above filter on the two columns `Func.refGene` and `ExonicFunc.refGene`. 
    * In `Func.refGene`, filter rows such that  *splicing* or *exonic* remains and 
    * in`ExonicFunc.refGene` column, *frameshift* or *nonsynonymous SNV* or *stopgain* or *stoploss* is included. 
  Compare to `corpasome.trio.chr22.snps.filtered.annovar.recessive.noncommon.exonsplicing`. 

```{r, solution = TRUE}
uncommon %>% 
  filter(Func.refGene %in% c("splicing", "exonic")) %>% 
  filter(ExonicFunc.refGene %in% c("frameshift", "nonsynonymous SNV", "stopgain","stoploss")) -> exonsplicing
```
  4. Now, we want to filter for the deleterious variants. Remove other patterns such that either *D* or *P* from each of the columns `ljb2_sift` and `ljb2_pp2hvar` remains. Compare to `corpasome.trio.chr22.snps.filtered.annovar.recessive.noncommon.exonsplicing.pred`. Could you also think of an explanation as to why we want to keep the deleterious variants ?
  
```{r, solution = TRUE}
exonsplicing %>% 
 filter( str_detect(ljb2_sift, "D|P")) %>% 
 filter( str_detect(ljb2_pp2hvar, "D|P")) -> cands

```
  
  
  5. Finally, we want to see how many of our predicted genes overlap with the candidategene list. Run an overlap between the predicted file (produced by the last question) and the `candidateGenesList` in `WEScourse/data/`.   Are there any matching genes ?
  
```{r, solution = TRUE}
cand_genes <- read_tsv("/Users/roland.krause/Workspace/isb101/NGS/WEScourse/data/candidateGenes/schizophrenia.cand.genes.ids", col_names = FALSE)
semi_join(cands, cand_genes, by=c(Gene.refGene = "X1"))
```

 6. Generate a bar plot that shows the numbers of variants in the individual filtering steps.

  
## *De novo* annotation analysis
  1. Load the file `corpasome.trio.chr22.denovogear.annovar_input.hg19_multianno.txt` into your R session.
  
```{r, solution = TRUE}
denovo <- read_tsv("/Users/roland.krause/Workspace/isb101/NGS/WEScourse/corpasome/denovo/corpasome.trio.chr22.denovogear.annovar_input.hg19_multianno.txt", na="")
```
  
  2. Filter the columns `esp6500si_ea` and `1000g2012apr_eur` for values less or equal to $0.00001$ and compare to 'corpasome.trio.chr22.denovogear.annovar_input.hg19_multianno.txt.noncommon`.
```{r, solution = TRUE}
cut_at <- 0.00001

denovo %>% 
  mutate(esp = replace_na(esp6500si_ea, 0)) %>% 
  mutate(tg = replace_na(`1000g2012apr_eur`, 0)) %>% 
  filter(esp <= cut_at ,
          tg <= cut_at ) -> uncommon
```
  
  3. From the previous file, remove the variants already present in dbSNP listed in the column `snp138`.
  
```{r, solution = TRUE}
uncommon %>% filter(is.na(snp138)) -> dbsnp

```

  4. From the file `corpasome.trio.chr22.denovogear.annovar_input.hg19_multianno.txt.noncommon..notdbSNP`, alter two columns `Func.refGene` and `ExonicFunc.refGene`. In `Func.refGene`, filter rows such that  *splicing* or *exonic;splicing* remains and in `ExonicFunc.refGene` column, *frameshift* or *nonsynonymous* or *stopgain* or *stoploss* is included. Save all of these changes into a variable named `denovogear.exonsplicing`
  
```{r, solution = TRUE}
dbsnp %>% 
  filter(str_detect(Func.refGene, "exonic")) %>% 
  filter(str_detect(ExonicFunc.refGene, c("nonsynonymous|frameshift|stopgain|stoploss"))) -> denovogear.exonsplicing
  
```
  
  5. Now, we want to filter for the deleterious variants. Remove other patterns such that only  `D` from each of the columns of ljb2_sift,ljb2_pp2hvar remains. Save this into a new variable called to denovogear.pred
```{r, solution = TRUE}
denovogear.exonsplicing %>% 
 filter(str_detect(ljb2_sift, "D")) %>% 
 filter(str_detect(ljb2_pp2hvar, "D")) -> denovogear.pred
```
  
  6. Same as before, we want to see how many of our predicted genes overlap with the candidate gene list. Run an overlap between the predicted file (produced by the last question) and the candidateGenesList. Are there any matching genes ? If yes, which ones.
  
```{r, solution = TRUE}
semi_join(denovogear.pred, cand_genes, by=c(Gene.refGene = "X1"))
```

 7. Generate a bar plot that shows the numbers of variants in the individual filtering steps.

### Quality control with FastQC

Try reading the fastqc_data.txt for one of the samples. Reading in this file is a bit complicated, make sure you have read the help messages for `read_tsv` and the related functions. Do not edit the file manually!
