---
title: "Access to databases"
author: "Roland Krause"
date: "2019-03-26"
output:
  ioslides_presentation:
    logo: logo-lcsb.png
    smaller: yes
  beamer_presentation: default
highlighter: highlight.js
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)

```

## Objectives|Discovering and access of biological information

1. Browsing, identify structure of data
2. Access data programmatically - e.g. using R, wget or other 
via an Application Program Interface (API)

## Standard text formats for data
> -  ### Tab-separated text files ('''.tab/.tsv''')
> -  ### eXtensible Markup Language (XML)
> -  ### JSON

## How data is organized|Large institutions to host data
- NCBI - National Center for Biotechnology Information
- EBI - European Institute for Bioinformatics

## Some databases
 * [NCBI Gene](http://www.ncbi.nlm.nih.gov/gene) - Gene sequences
 * [Uniprot](http://www.uniprot.org) - Protein sequences
 * [STRING](www.string-dg.org) - Protein-protein interactions
 * [OMIM](https://www.ncbi.nlm.nih.gov/omim/) - Mendelian diseases and genes   

## Genome browsers

- UCSC Genome Browser - [genome.ucsc.edu](genome.ucsc.edu)
- ENSEMBL - [www.ensembl.org](www.ensembl.org)


## Curated resources | Uniprot and Refseq

[Uniprot](http://www.uniprot.org)


# Working with biological database

## Advice for your projects

#### Looking up data on the web -- not a reliable resource for your thesis
 $\to$ Save your queries
 
#### Occasional look-up in your database or from a linked resource
 $\to$ Programmatic access

#### Full resource, e.g. when computing statistics
 $\to$ Download full resource, denote version, date and URL or (ideally Digital Object Identifier) 
 Script but ensure that it's not automatically triggered!  
 Your institute's bioinformatics facility might already hold copy

#### Mirror installation of larger resourcer for bigger project

## Downloading data in R 

Ensure you are not downloading files on every execution of your script.
```{r download}
 destfile="./hordeum_variants.txt.gz" 
 fileURL <- "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/variants/hordeum_vulgare_variation.txt.gz"

 if (!file.exists(destfile)) {
    download.file(fileURL ,destfile,method="auto") }
 

```


# Programmatic access to databases 

## BioMart installation

```{r, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("biomaRt")

```


```{r}
library("biomaRt")
listMarts(host="www.ensembl.org")
```


## Using ENSEMBL Mart

```{r }
ensembl.mart = useMart(biomart="ENSEMBL_MART_ENSEMBL", 
                       host="www.ensembl.org")
head(listDatasets(ensembl.mart))
```

## Using a specific database
```{r}
human.genes.mart = useDataset(ensembl.mart, dataset="hsapiens_gene_ensembl")
head(listAttributes(human.genes.mart))

```

## Get data from the database

```{r}
chr21_genes <- getBM(attributes=c('ensembl_gene_id',
'ensembl_transcript_id','hgnc_symbol','chromosome_name','start_position','end_position'), filters =
'chromosome_name', values ="21", mart = human.genes.mart)

head(chr21_genes)
```

## Links between databases


```{r, eval=FALSE}
hgnc_swissprot <- getBM(attributes=c('ensembl_gene_id','ensembl_transcript_id','hgnc_symbol','hgnc_id','uniprot_swissprot'),filters = 'ensembl_gene_id', values = 'ENSG00000139618', mart = human.genes.mart)
head(hgnc_swissprot)
```


## MutationMapper interface

```{r}
library(cgdsr)
# Create CGDS object
mycgds = CGDS("http://www.cbioportal.org/")


# Get list of cancer studies at server
getCancerStudies(mycgds)

```

## MutationMapper interface
```{r}

# Get available case lists (collection of samples) for a given cancer study
mycancerstudy = getCancerStudies(mycgds)[2,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]

# Get available genetic profiles
mygeneticprofile = getGeneticProfiles(mycgds,mycancerstudy)[4,1]

# Get data slices for a specified list of genes, genetic profile and case list
getProfileData(mycgds,c('BRCA1','BRCA2'),mygeneticprofile,mycaselist)

# Get clinical data for the case list
myclinicaldata = getClinicalData(mycgds,mycaselist)

# documentation

```

## MutationMapper interface

```{r}


plot(mycgds, "gbm_tcga", "NF1", c("gbm_tcga_gistic","gbm_tcga_mrna"), "gbm_tcga_all")
```



