---
title: "Programmatic access to the ENSEMBL database"
author: "Roland Krause"
date:  "`r Sys.Date()`"
output:
  ioslides_presentation:
    logo: logo-lcsb.png
    smaller: yes
  beamer_presentation: default
highlighter: highlight.js
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)

```

## Learning Objectives 

### Discovering and access of biological information {.box-10 .offset-1 .bg-red .icon}

![](img/00/kt.png)

 1. Browsing and  identifying structures of data
 2. Access data programmatically 
    * with R 
    * or shell tools (`wget`  `curl`)
    * via an Application Program Interface (API)

## Formats for data exchange {.build}

### Standards in 2020 {.box-6 .offset-3 .bg-blue}

> -  ### Tab-separated text files ('''.tab/.tsv''')
> -  ### eXtensible Markup Language (XML)
> -  ### JSON

## How data is organized|Large institutions to hosting

### US-based {.box-4 .offset-1 .bg-blue}
 - UCSC Genome Browser - [genome.ucsc.edu](genome.ucsc.edu)
 - [Refseq](https://www.ncbi.nlm.nih.gov/refseq/)
 - National Center for Biotechnoloy Information [NCBI](https://www.ncbi.nlm.nih.gov/)

### Europe {.box-4 .offset-2 .bg-green .stretch}
 - ENSEMBL - [www.ensembl.org](http://www.ensembl.org)
 - [Uniprot](http://www.uniprot.org)
 - EBI - European Institute for Bioinformatics


# Working with biological databases

## Advice for your projects {.build}

### Looking up data on the web {.bg-green .box-6 }
 * Not a reliable resource for your thesis
 * Use to discover structure of a data set or hosting site
 
 
 $\to$ Save your queries
 
### Repeated look-ups in a database  {.bg-yellow .box-6 .stretch}
 * Exploring several genes in a consistent fashion
 
 
  $\to$ Programmatic access reommended

### Genome-wide computations {.bg-blue .box-8 }
 * Programmatic access required 
 * Ensure that installations are not automatically triggered!  
 * Your institute's bioinformatics facility might already hold a copy of a resource.
 * Mirror installation of larger resourcer are big ($>$ 3Tb for ENSEMBL)
  
 $\to$ Download full resource, denote version, date and URL or (ideally Digital Object Identifier) 
 
### {.box-4 }
 ![](img/05/hard-disk.jpg)


## Downloading data in R 

Ensure you are not downloading files on every execution of your script.
```{r download}
 destfile="./hordeum_variants.txt.gz" 
 fileURL <- "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/variants/hordeum_vulgare_variation.txt.gz"

 if (!file.exists(destfile)) {
    download.file(fileURL ,destfile,method="auto") }
 

```


# Programmatic access to databases 

## BioMart installation  {.col-8 .offset-2 .box .bg-blue}

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")
```


```{r}
library("biomaRt")
listMarts(host="www.ensembl.org")
```


## Using ENSEMBL Mart {.build}

### BioMart on the web {.bg-green .box-8 .offset-2}
 * [BioMart](https://www.ensembl.org/info/data/biomart)

### BioMart from the API  {.bg-blue .box-8 .offset-2}
```{r }
ensembl.mart = useMart(biomart="ENSEMBL_MART_ENSEMBL", 
                       host="www.ensembl.org")
head(listDatasets(ensembl.mart))
```

## Using a specific database
```{r}
human.genes.mart = useDataset(ensembl.mart, dataset="hsapiens_gene_ensembl")
head(listAttributes(human.genes.mart))

```

## Get data from the database

```{r}
chr21_genes <- getBM(attributes=c('ensembl_gene_id',
'ensembl_transcript_id','hgnc_symbol','chromosome_name','start_position','end_position'), filters =
'chromosome_name', values ="21", mart = human.genes.mart)

head(chr21_genes)
```

## Links between databases


```{r, eval=FALSE}
hgnc_swissprot <- getBM(attributes=c('ensembl_gene_id','ensembl_transcript_id','hgnc_symbol','hgnc_id','uniprot_swissprot'),filters = 'ensembl_gene_id', values = 'ENSG00000139618', mart = human.genes.mart)
head(hgnc_swissprot)
```


## MutationMapper interface

```{r}
library(cgdsr)
# Create CGDS object
mycgds = CGDS("http://www.cbioportal.org/")


# Get list of cancer studies at server
getCancerStudies(mycgds)

```

## MutationMapper interface
```{r}

# Get available case lists (collection of samples) for a given cancer study
mycancerstudy = getCancerStudies(mycgds)[2,1]
mycaselist = getCaseLists(mycgds,mycancerstudy)[1,1]

# Get available genetic profiles
mygeneticprofile = getGeneticProfiles(mycgds,mycancerstudy)[4,1]

# Get data slices for a specified list of genes, genetic profile and case list
getProfileData(mycgds,c('BRCA1','BRCA2'),mygeneticprofile,mycaselist)

# Get clinical data for the case list
myclinicaldata = getClinicalData(mycgds,mycaselist)

# documentation

```

## MutationMapper interface

```{r}


plot(mycgds, "gbm_tcga", "NF1", c("gbm_tcga_gistic","gbm_tcga_mrna"), "gbm_tcga_all")
```



