---
title: "IT101 - Working with computers"
author: "Roland Krause"
date: "`r Sys.Date()`"
smaller: true
logo: "../../trainingplan/img/r3-logo-square-01.png"
output:
  iosp::ioslides_plus:
    footer: "IT101"
    box_colours:
      bg-my_colour: "slateblue2"
      bg-epipgx: [ "gold", "darkmagenta"]
      bg-unilu: ["#ee2a28", "darkred"]
      bg-mechepi: [ "lightskyblue", "darkorange"]
      bg-ilae: ["darkmagenta", "darkorchid1"]
      bg-my_colour2: ["tomato3", "tomato2"]
      bg-intro: ["tomato3", "darkorange"]
      bg-my_colour3: ["springgreen4", header_text: "#FFD700"]
      bg-lcsb: ["darkred", "lightblue1"]
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#library(ggbeeswarm)
library(tidyverse)
library(knitr)
```


```{css, echo = FALSE}
.box-body {
  font-size: 16pt;
  }

.box-body > .chunk > h3 {
  font-size: 16pt;
  line-height: 24pt;
}

.ss {
  font-size: 16pt;
  
}

.p {
 font-size: 16pt;
}


.icon-small img {
  float: right;
  margin-left: 20px;
  max-height: 75px !important;
}
```

## Scientific data management by typical scientists

### Dr Bruce Banner ... {.box-6  .bg-green .stretch .build}

 ... has 7 PhDs  
 
 ... no publication record
 
 ... stores all research data in a database  
 
 ... experiments fully reproducible
 
### {.box-4 .offset-1 }
![](img/00/hulk.png)

## Introduction {.build}

### Learning objectives{.box-6 .bg-green}
 * Reproduciblity in the research data life cycle
 * How to manage your data
 * How to look and analyse your data
 * Solving issues with computers
 * Get to know support from Bioinformatics core  

   
### Pertains to practically all people at LCSB {.box-6  .bg-lcsb .stretch .build}
 * Scientists
 * PhD candidates
 * Technicians
 * Administrators



## Typical flow of data {.smaller .build}

###  Source data ➡️ {.col-3 .box .bg-green .outline .smaller}
 * Experimental results 
 * Large data sets
 * Manually collected data
 * External
 
###  Intermediate➡️ {.col-3 .box .bg-blue .stretch} 
 * Derived data
 * *Tidy* data 
 * Curated sets

###  Analysis ➡️  {.col-3 .box .bg-yellow .icon .stretch}
 * Exploratory
 * Model building
 * Hypothesis testing

###  Manuscript{.box-3 .warning .bg-lcsb  .stretch}
Reproduce

 * all numbers
 * graphs and figures
  
### Preserve {.box-3 .bg-green .stretch .outline}
Version data sets.

### Reproduce {.box-3 .bg-blue .stretch .outline}
Automate your *builds*. Use workflow tools (e.g. `Snakemake`)

### Trace  {.box-3 .bg-yellow .stretch .outline}
Multiple iterations. Code versioning (`git` repositories)

### Track {.col-3 .box .bg-lcsb .stretch .outline}
Through multiple versions. Integrated with analysis (e.g. `RMarkdown` documents)

# All data processing is computation.

## Why is your workflow not reproducible?

### Lack of *provenance*  {.box-7  .bg-intro .smaller .stretch}
 * Input data from downloaded from “some website” 
 * Copy & paste operations
 * Manual text entry
 * Analysis not coded
 * Intermediate and final data **not** generated by deterministic processes

## Learning to code workflows and analysis

### Excel alone {.box-6 .bg-green} 
 + Is great for looking at data.
 + Data entry is fast.
 + Analysis flow is hidden and not in focus. 

### Coding {.box-6 .bg-blue}
 + Is great for controling analysis workflow
 + Data is hidden.
 + Flow is visible.
 
### Develop data science skills {.box-12 .bg-lcsb}
 * Develop good data management and analysis *habits*.
 * Start coding your analysis within Excel.
 * Make yourself familiar with a statistics environment such as R, Python or Matlab
 * No need to learn a high level programming language such as C++ or Java.


## So you're not a programmer{ .outline .build}

### {.box-12}
![](../../trainingplan/img/notaprogrammer.png)

### A *real programmer* will not understand your research!  {.box-12 .bg-red }


# Data first.
## Receiving and sending data 

### E-mail is not for data transfer {.box-6  .bg-red .smaller .stretch .icon-small}
 * Avoid transfer of any data by e-mail
 * E-mail is a poor repository
 * Data can be read on passage
 * Use a file sharing service 

 * Special precautions for sensitive data
   ![](img/warning.png) 

### Exchanging data {.box-6  .bg-blue .smaller .stretch  .icon-small} 
 * Share on Atlas server 
 * dropit.uni.lu service (SIU)
 * ownCloud share (LCSB - BioCore)
 * Commercial vendors 
    + Not generally advised. Collaborators. 
    + Dropbox, Google Drive, OneDrive, iCloud



### Advice  {.box-12  .bg-yellow .smaller .stretch .icon-small}
 * Use institutional media for storage of all data
 * Disable write access to relevant source data
 * Clarify and separate source and intermediate data

![](img/tip.png)

## Filenames | Conventions {.build}

### General principles {.box-4 .bg-yellow}
 * Machine readable
 * Human readable
 * Plays well with default ordering 
 
### Separators {.box-4  .bg-lcsb .smaller .stretch}
 * No spaces
 * Underscore to separate
 * Hyphen to combine

### Date format {.box-4  .bg-yellow .smaller .stretch}
ISO8601

`2018-12-03`

`2018-12-06_1700`

### Examples {.box-6  .bg-green .ss}
> - `PhD-project-Jan19 alldata_final.foo`\
> - `Finacial detailes BIocore 19/11/12.xls`\
>- `ATACseq1Londonmapped.bam`\

### Examples {.box-6  .bg-blue .ss}
> - `Iris-setosa_samples_1927-05-12.csv`\
> - `PI102_Mouse12_EEG_2018-11-03_1245.tsv`\
> - `Bioinfiniti_FullProposal_2018-11-15_1655.docx`


### From Jenny Bryan by CC-BY  {.outline .box-12 .smaller}
[https://speakerdeck.com/jennybc/how-to-name-files](https://speakerdeck.com/jennybc/how-to-name-files)


 
 
# Backup

## Taking care of your back-up {.build}

### No automatic back-up {.box-7 .bg-unilu .stretch}
 * You have to take care of your own back-up!  
 * Research data, particularly sensitive data should be in a single source location.
 * Don't *work* of your back-up copy!
 
### 3-2-1 Back-up {.box-5   .bg-yellow .stretch}
 * Three copies
 * Two different media
 * One off-site copy

### Examples {.box-12 .bg-lcsb}

### Local data {.box-4 .bg-green .stretch }
 * Working documents on your laptop
 * Online share (owncloud)
 * Copy on an external hard drive

### Group data {.box-4 .bg-blue .stretch}
 * Research data generated in group
 * Back-up by central IT
 * Georesilient copy

### Sensitive data {.box-4 .bg-yellow .stretch }
 * Data set on virtual machine
 * Encryption
 * See details by Pinar later



## Code versioning

### `git` {.box-6 .bg-green}
 * Current standard for code versioning
 * Maintain versions of your code as it develops
 * Local system, which does not require an online repository
 * Repositories allow distributed development 
 
### git@lcsb {.box-6 .bg-lcsb .stretch}
 * Recommended, supported repository
 * Allows tracking of issues
 * Ready for continous integration - code checked on commits to the repository.
 * [https://git-r3lab.uni.lu](https://git-r3lab.uni.lu)

### Use at LCSB {.box-8  .offset-2 .bg-yellow .icon-small}
 * All analysis code should be in a repository
   * Minimally at submission of a manuscript
   * Better daily
   * Even better "analysis chunkwise" 
   ![](img/00/tip.png)



# Visualization

## Dynamite graphs {.build}

```{r include = FALSE}
iris %>%
  sample_n(50) %>%
  group_by(Species) %>%
  summarize(mean_sl = mean(Sepal.Length),
            sd_sl = sd(Sepal.Length),
            n = n(),
            se_sl = sd_sl / sqrt(n())
            ) -> p
g_se <- ggplot(p, aes(x = Species, y = mean_sl, fill = Species)) +
          geom_col() 

```

### Mean and Standard Error {.box-6 .bg-blue}
```{r, echo=FALSE, fig.width=6}
g_se + 
  geom_errorbar(aes(ymin = mean_sl - se_sl, 
                    ymax = mean_sl + se_sl)) 
```

### Mean and Standard Deviation  {.box-6 .bg-blue}
```{r echo=FALSE, fig.width=6}
g_se + 
geom_errorbar(aes(ymin = mean_sl - sd_sl, 
                  ymax = mean_sl + sd_sl)) 
```

## Better display options {.build}

```{r echo=FALSE }
ggplot(sample_n(iris,80),  
       aes(x = Species, y = Sepal.Length, fill=Species )) + 
  theme(legend.position="bottom") -> g
```

### Box plot {.box-4 .bg-blue}
```{r  echo=FALSE, fig.width=4, height = 6}
g + geom_boxplot() 
```

### Violin plots {.box-4 .bg-blue}
```{r,  echo=FALSE, fig.width=4,  height = 6}
g  + geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) 
```


### Jittered display{.box-4 .bg-blue}
```{r, echo=FALSE, fig.width=4, height = 6}
g + aes(colour=Species) + geom_jitter()
```




## Show me the data! {.build}

### Bar chart with data {.box-8 .bg-blue}
```{r showbar, echo=FALSE, fig.height=4}
ggplot(data = iris %>% group_by(Species) %>% summarize_all(mean)) + 
  aes(x = Species, y = Sepal.Length, fill = Species) +
  geom_col(alpha =0.2) + 
  geom_jitter(data= iris, aes(x=Species, y = Sepal.Length))
```

### Violin plot{.box-4 .bg-blue .stretch}
```{r, error = TRUE, fig.height=4, echo=FALSE}
g + 
  geom_violin() +
  geom_jitter(height = 0, width = 0.2)
```

## plot your data! | [Anscombe](https://d2f99xq7vri1nk.cloudfront.net/Anscombe_1_0_0.png) ** 10 {.vs2}

> never trust summary statistics alone; always visualize your data | Alberto Cairo {.bg-grgold}

![](https://d2f99xq7vri1nk.cloudfront.net/DinoSequentialSmaller.gif)

## Problem solving

### A guide for solving computing issues {.box-8 .offset-2 .bg-yellow .outline .icon-small}

![](img/00/tip.png)

 1. Express the problem
 2. Search for help
 3. Ask an expert

## Rubberducking

### Tell it to your rubberduck! {.box-4 .offset-6}
![](../../trainingplan/img/1000px-Rubber_duck.svg.png)

## Google for help {.build}

### Even decent programmers use Google {.box-10 .bg-yellow .offset-1}
Why shouldn't I?

### General flow {.box-10 .bg-green .offset-1}
  * Read help pages and the official documentation well before turning to Google
  * Use stack exchange, forums and related resources carefully
  + Answers for Excel questions are worst
  * Write down what you want to achieve before turning to Google


## Write to the Oracle {.build}

### The Oracle gives the precise answer to your problems {.box-10 .offset-1 .bg-yellow}
 * You have to submit the problem in writing
 * The Oracle answers a questions only once or if it finds the problem interesting
 * If you supply a trivial problem, it will stop answering
 
### Available Oracles  {.box-10 .offset-1 .bg-lcsb}
 * Service Now @ [service.uni.lu](service.uni.lu) (Uni and LCSB helpdesk)
 * Stack Overflow and other online sites
 * Local experts


## Your $R^3$ contacts {.smaller}

### Christophe Trefois {.box-3 .bg-lcsb .stretch}
![](img/00/christophe_trefois.png)
$R^3$ Coordination

### Laurent Heirendt  {.box-3 .bg-lcsb .stretch .icon}
![](img/00/laurent_heirendt.png)
git, CI


### Pinar Alper {.box-3 .bg-lcsb}
![](img/00/pinar_alper.png)
Data steward

### Sarah Peter {.box-3 .bg-lcsb .stretch}
![](img/00/sarah_peter.png)
High Performance Computing (HPC)


### Wei Gu {.box-3 .bg-lcsb}
![](img/00/wei_gu.png)
 $R^3$ Core

### Venkata Satagopam  {.box-3 .bg-lcsb .smaller .stretch }
![](img/00/venkata_satagopam.png)
$R^3$ Core

### Yohan Jarosz {.box-3 .bg-lcsb .stretch}
![](img/00/yohan_jarosz.png)
$R^3$ Core
 
### Reinhard Schneider {.box-3 .bg-lcsb}
![](img/00/reinhard_schneider.png)
Head of Bioinformatics Core

## Additional acknowledgements

### Thank you ... {.box-6 .offset-3 .bg-yellow}
for your attention!


###  {.box-10 .offset-1 .small .bg-blue}
 * Eric Koncina (LSRU, `iosp`, rendering this presentation)
 * Aurélien Ginohac (LSRU)

